File: examples/go/ttl-caching/kv_operations.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"os"

	"github.com/joho/godotenv"
)

func main() {
	// Load environment variables from the .env file
	err := godotenv.Load()
	if err != nil {
		log.Fatal(fmt.Sprintf("✗ Failed to load environment variables: %v", err))
		os.Exit(1)
	}

	var baseURL = os.Getenv("API_BASE_URL")
	var apiKey = os.Getenv("API_BASE_KEY")

	// Get authentication token
	token := os.Getenv("AUTH_TOKEN")
	if token == "" {
		tokenResp, err := http.Post(baseURL+"/api/auth/token", "application/json",
			bytes.NewBuffer([]byte(fmt.Sprintf(`{"api_key":"%s"}`, apiKey))))
		if err != nil {
			log.Fatal(err)
		}
		defer tokenResp.Body.Close()

		var tokenData map[string]interface{}
		json.NewDecoder(tokenResp.Body).Decode(&tokenData)
		token = tokenData["token"].(string)
	}

	fmt.Println("✓ Authentication successful")

	client := &http.Client{}

	// Example 1: Set a key-value pair
	fmt.Println("\n=== KV Set ===")
	kvData := map[string]interface{}{
		"value": map[string]interface{}{
			"userId":   123,
			"username": "john_doe",
		},
	}
	kvJSON, _ := json.Marshal(kvData)

	req, _ := http.NewRequest("POST", baseURL+"/api/kv/set/session:user123", bytes.NewBuffer(kvJSON))
	req.Header.Set("Authorization", "Bearer "+token)
	req.Header.Set("Content-Type", "application/json")

	resp, _ := client.Do(req)
	resp.Body.Close()
	fmt.Println("✓ Set key: session:user123")

	// Example 2: Get a key-value pair
	fmt.Println("\n=== KV Get ===")
	req, _ = http.NewRequest("GET", baseURL+"/api/kv/get/session:user123", nil)
	req.Header.Set("Authorization", "Bearer "+token)

	resp, _ = client.Do(req)
	var getValue map[string]interface{}
	json.NewDecoder(resp.Body).Decode(&getValue)
	resp.Body.Close()
	fmt.Println("Retrieved value:", getValue["value"])

	// Example 3: Set multiple keys
	fmt.Println("\n=== Set Multiple Keys ===")
	keys := []string{"cache:product:1", "cache:product:2", "cache:product:3"}
	for i, key := range keys {
		data := map[string]interface{}{
			"value": map[string]interface{}{
				"name":  fmt.Sprintf("Product %d", i+1),
				"price": 29.99 + float64(i*10),
			},
		}
		dataJSON, _ := json.Marshal(data)

		req, _ := http.NewRequest("POST", baseURL+"/api/kv/set/"+key, bytes.NewBuffer(dataJSON))
		req.Header.Set("Authorization", "Bearer "+token)
		req.Header.Set("Content-Type", "application/json")

		resp, _ := client.Do(req)
		resp.Body.Close()
	}
	fmt.Printf("✓ Set %d keys\n", len(keys))

	// Example 4: Get multiple keys
	fmt.Println("\n=== Get Multiple Keys ===")
	for _, key := range keys {
		req, _ := http.NewRequest("GET", baseURL+"/api/kv/get/"+key, nil)
		req.Header.Set("Authorization", "Bearer "+token)

		resp, _ := client.Do(req)
		var result map[string]interface{}
		json.NewDecoder(resp.Body).Decode(&result)
		resp.Body.Close()
		fmt.Printf("%s: %v\n", key, result["value"])
	}

	// Example 5: Delete a key
	fmt.Println("\n=== KV Delete ===")
	req, _ = http.NewRequest("DELETE", baseURL+"/api/kv/delete/session:user123", nil)
	req.Header.Set("Authorization", "Bearer "+token)

	resp, _ = client.Do(req)
	resp.Body.Close()
	fmt.Println("✓ Deleted key: session:user123")

	// Verify deletion
	req, _ = http.NewRequest("GET", baseURL+"/api/kv/get/session:user123", nil)
	req.Header.Set("Authorization", "Bearer "+token)

	verifyResp, _ := client.Do(req)
	verifyResp.Body.Close()
	if verifyResp.StatusCode == 404 {
		fmt.Println("✓ Verified: Key successfully deleted (not found)")
	} else {
		fmt.Println("✗ Warning: Key still exists after delete!")
	}

	// Example 6: Delete multiple keys
	fmt.Println("\n=== Delete Multiple Keys ===")
	for _, key := range keys {
		req, _ := http.NewRequest("DELETE", baseURL+"/api/kv/delete/"+key, nil)
		req.Header.Set("Authorization", "Bearer "+token)

		resp, _ := client.Do(req)
		resp.Body.Close()
	}
	fmt.Printf("✓ Deleted %d keys\n", len(keys))

	fmt.Println("\n✓ All KV operations completed successfully")
}

File: examples/rust/examples/kv_operations.rs
use serde_json::{json, Value};
use std::env;

const SESSION_KEY: &str = "kv_operations:direct:rs:session:user123";
const PRODUCT_KEYS: [&str; 3] = [
    "kv_operations:direct:rs:cache:product:1",
    "kv_operations:direct:rs:cache:product:2",
    "kv_operations:direct:rs:cache:product:3",
];

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    dotenv::dotenv().ok();

    let base_url = env::var("API_BASE_URL").unwrap_or_else(|_| "http://localhost:8080".to_string());
    let api_key =
        env::var("API_BASE_KEY").unwrap_or_else(|_| "a-test-api-key-from-ekodb".to_string());

    // Step 1: Get authentication token
    let client = reqwest::Client::new();
    let token_response = client
        .post(&format!("{}/api/auth/token", base_url))
        .json(&json!({ "api_key": api_key }))
        .send()
        .await?
        .error_for_status()?;

    let token_data: Value = token_response.json().await?;
    let token = token_data["token"].as_str().unwrap();
    println!("✓ Authentication successful");

    cleanup_owned_keys(&client, &base_url, token).await?;

    let operation_result = run_examples(&client, &base_url, token).await;
    let cleanup_result = cleanup_owned_keys(&client, &base_url, token).await;

    match (operation_result, cleanup_result) {
        (Ok(()), Ok(())) => {
            println!("\n✓ All KV operations completed successfully");
            Ok(())
        }
        (Err(operation_error), Ok(())) => Err(operation_error),
        (Ok(()), Err(cleanup_error)) => Err(cleanup_error),
        (Err(operation_error), Err(cleanup_error)) => Err(std::io::Error::other(format!(
            "{operation_error}; cleanup also failed: {cleanup_error}"
        ))
        .into()),
    }
}

async fn run_examples(
    client: &reqwest::Client,
    base_url: &str,
    token: &str,
) -> Result<(), Box<dyn std::error::Error>> {
    // Example 1: Set a key-value pair
    println!("\n=== KV Set ===");
    let kv_data = json!({
        "value": {
            "userId": 123,
            "username": "john_doe"
        }
    });

    client
        .post(&format!("{}/api/kv/set/{}", base_url, SESSION_KEY))
        .header("Authorization", format!("Bearer {}", token))
        .json(&kv_data)
        .send()
        .await?
        .error_for_status()?;

    println!("✓ Set key: session:user123");

    // Example 2: Get a key-value pair
    println!("\n=== KV Get ===");
    let get_response = client
        .get(&format!("{}/api/kv/get/{}", base_url, SESSION_KEY))
        .header("Authorization", format!("Bearer {}", token))
        .send()
        .await?
        .error_for_status()?;

    let get_value: Value = get_response.json().await?;
    println!("Retrieved value: {:?}", get_value["value"]);

    // Example 3: Set multiple keys
    println!("\n=== Set Multiple Keys ===");
    let keys = PRODUCT_KEYS;

    for (i, key) in keys.iter().enumerate() {
        let data = json!({
            "value": {
                "name": format!("Product {}", i + 1),
                "price": 29.99 + (i as f64 * 10.0)
            }
        });

        client
            .post(&format!("{}/api/kv/set/{}", base_url, key))
            .header("Authorization", format!("Bearer {}", token))
            .json(&data)
            .send()
            .await?
            .error_for_status()?;
    }
    println!("✓ Set {} keys", keys.len());

    // Example 4: Get multiple keys
    println!("\n=== Get Multiple Keys ===");
    for key in &keys {
        let response = client
            .get(&format!("{}/api/kv/get/{}", base_url, key))
            .header("Authorization", format!("Bearer {}", token))
            .send()
            .await?
            .error_for_status()?;

        let result: Value = response.json().await?;
        println!("{}: {:?}", key, result["value"]);
    }

    // Example 5: Delete a key
    println!("\n=== KV Delete ===");
    client
        .delete(&format!("{}/api/kv/delete/{}", base_url, SESSION_KEY))
        .header("Authorization", format!("Bearer {}", token))
        .send()
        .await?
        .error_for_status()?;

    println!("✓ Deleted key: session:user123");

    // Verify deletion
    let verify_response = client
        .get(&format!("{}/api/kv/get/{}", base_url, SESSION_KEY))
        .header("Authorization", format!("Bearer {}", token))
        .send()
        .await?;

    if verify_response.status() == 404 {
        println!("✓ Verified: Key successfully deleted (not found)");
    } else {
        println!("✗ Warning: Key still exists after delete!");
        return Err(std::io::Error::other("session key still exists after deletion").into());
    }

    // Example 6: Delete multiple keys
    println!("\n=== Delete Multiple Keys ===");
    for key in &keys {
        client
            .delete(&format!("{}/api/kv/delete/{}", base_url, key))
            .header("Authorization", format!("Bearer {}", token))
            .send()
            .await?
            .error_for_status()?;
    }
    println!("✓ Deleted {} keys", keys.len());

    Ok(())
}

async fn cleanup_owned_keys(
    client: &reqwest::Client,
    base_url: &str,
    token: &str,
) -> Result<(), Box<dyn std::error::Error>> {
    let mut errors = Vec::new();
    for key in std::iter::once(SESSION_KEY).chain(PRODUCT_KEYS) {
        match client
            .delete(format!("{base_url}/api/kv/delete/{key}"))
            .header("Authorization", format!("Bearer {token}"))
            .send()
            .await
        {
            Ok(response)
                if response.status().is_success()
                    || response.status() == reqwest::StatusCode::NOT_FOUND => {}
            Ok(response) => errors.push(format!("{key}: HTTP {}", response.status())),
            Err(error) => errors.push(format!("{key}: {error}")),
        }
    }

    if errors.is_empty() {
        Ok(())
    } else {
        Err(std::io::Error::other(format!(
            "failed to clean owned KV keys: {}",
            errors.join("; ")
        ))
        .into())
    }
}
